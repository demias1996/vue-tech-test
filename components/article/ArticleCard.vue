<template>
    <b-card
        :title="article.title"
        :sub-title="article.author"
        :img-src="article.image"
        :img-alt="articleImageDescription"
        img-top
        tag="article"
        class="mb-4"
        body-class="d-flex flex-column"
    >
        <b-card-text>
            {{ article.intro }}
        </b-card-text>

        <nuxt-link class="article-link mt-auto" :to="articleRoute">
            <b-button class="w-100" variant="primary">
                Open article
            </b-button>
        </nuxt-link>
    </b-card>
</template>

<script>
    export default {
        name: 'article-card',

        props: {
            article: {
                type: Object,
                required: true,
            },
        },

        computed: {
            /**
             * Returns description with some details about article for image.
             *
             * @returns {string}
             */
            articleImageDescription() {
                const { author, title } = this.article;

                return `Article from ${ author } with title ${ title }`;
            },

            /**
             * Returns _article route with appropriate id.
             *
             * @returns {string}
             */
            articleRoute() {
                return `/article/${ this.article.id }`;
            },
        },
    };
</script>
