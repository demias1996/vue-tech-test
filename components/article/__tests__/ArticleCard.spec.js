import { shallowMount } from '@vue/test-utils';
import ArticleCard from '@/components/article/ArticleCard';
import { ArticleFixture } from '@/fixtures/ArticleFixture';
import { BCard, BCardText } from 'bootstrap-vue'

/**
 * Returns wrapper for component shallowly mounted with required props.
 *
 * @param {object} propsData
 * @returns {Wrapper<Vue>}
 */
const shallowMountArticleCard = (propsData = {}) =>
    shallowMount(ArticleCard, { propsData: {
        article: { ...ArticleFixture },
        ...propsData,
    } });

describe('ArticleCard', () => {
    test('is a Vue instance', () => {
        const wrapper = shallowMountArticleCard();
        expect(wrapper.vm).toBeTruthy();
    });

    test('is rendered correctly with required prop', () => {
        const wrapper = shallowMountArticleCard();
        expect(wrapper.element).toMatchSnapshot();
    });

    test('is article title filled in card', () => {
        const wrapper = shallowMountArticleCard();
        const cardElement = wrapper.findComponent(BCard);

        expect(cardElement.attributes().title).toBe(ArticleFixture.title);
    });

    test('is article author name filled in card', () => {
        const wrapper = shallowMountArticleCard();
        const cardElement = wrapper.findComponent(BCard);

        expect(cardElement.attributes().subtitle).toBe(ArticleFixture.author);
    });

    test('is article image source filled in card', () => {
        const wrapper = shallowMountArticleCard();
        const cardElement = wrapper.findComponent(BCard);

        expect(cardElement.attributes().imgsrc).toBe(ArticleFixture.image);
    });

    test('is article image description filled in card', () => {
        const wrapper = shallowMountArticleCard();
        const cardElement = wrapper.findComponent(BCard);
        const { author, title } = ArticleFixture;

        expect(cardElement.attributes().imgalt).toBe(`Article from ${ author } with title ${ title }`);
    });

    test('is article intro filled', () => {
        const wrapper = shallowMountArticleCard();
        const cardTextElement = wrapper.findComponent(BCardText);

        expect(cardTextElement.text()).toBe(ArticleFixture.intro);
    });

    test('is article route set', () => {
        const wrapper = shallowMountArticleCard();
        const linkElement = wrapper.find('.article-link');

        expect(linkElement.attributes().to).toBe(`/article/${ ArticleFixture.id }`);
    });
});
