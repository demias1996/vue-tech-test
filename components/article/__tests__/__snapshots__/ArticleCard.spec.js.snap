// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard is rendered correctly with required prop 1`] = `
<b-card-stub
  bodyclass="d-flex flex-column"
  bodytag="div"
  class="mb-4"
  footertag="div"
  headertag="div"
  imgalt="Article from Test author with title Article title"
  imgsrc="https://test-image-source"
  imgtop="true"
  subtitle="Test author"
  subtitletag="h6"
  subtitletextvariant="muted"
  tag="article"
  title="Article title"
  titletag="h4"
>
  <b-card-text-stub
    texttag="p"
  >
    
        Article intro
    
  </b-card-text-stub>
   
  <a
    class="article-link mt-auto"
    to="/article/1"
  >
    <b-button-stub
      class="w-100"
      tag="button"
      target="_self"
      type="button"
      variant="primary"
    >
      
            Open article
        
    </b-button-stub>
  </a>
</b-card-stub>
`;
